security/pkg/pki/ca/ca.go

```diff
@@ -210,23 +226,85 @@ func NewSelfSignedIstioCAOptions(ctx context.Context,
 	return caOpts, err
 }
 
-func loadSelfSignedCaSecret(client corev1.CoreV1Interface, namespace string, caCertName string, rootCertFile string, caOpts *IstioCAOptions) error {
+// loadCASecrets loads the original istio-ca-secret Secret from K8S.
+// To simplify - this secret should be used for both self-generated and user defined intermediary certs.
+// TODO: rename to loadRootsFromSecret.
+func loadCASecrets(client corev1.CoreV1Interface, namespace string, caCertName string, rootCertFile string, caOpts *IstioCAOptions) error {
 	caSecret, err := client.Secrets(namespace).Get(context.TODO(), caCertName, metav1.GetOptions{})
 	if err == nil {
-		pkiCaLog.Infof("Load signing key and cert from existing secret %s/%s", caSecret.Namespace, caSecret.Name)
-		rootCerts, err := util.AppendRootCerts(caSecret.Data[CACertFile], rootCertFile)
+		// Original simple case - self-generate root, no intermediaries.
+		rootCertData := caSecret.Data[CACertFile]
+
+		// The file is typically the mounted ca - if the roots are present, they are added.
+		// Not sure if this is intentional or accidental.
+		rootCerts, err := util.AppendRootCerts(rootCertData, rootCertFile)
 		if err != nil {
 			return fmt.Errorf("failed to append root certificates (%v)", err)
 		}
-		if caOpts.KeyCertBundle, err = util.NewVerifiedKeyCertBundleFromPem(caSecret.Data[CACertFile],
-			caSecret.Data[CAPrivateKeyFile], nil, rootCerts); err != nil {
+
+		privData := caSecret.Data[CAPrivateKeyFile]
+		crtData := caSecret.Data[CACertFile]
+
+		if caOpts.KeyCertBundle, err = util.NewVerifiedKeyCertBundleFromPem(crtData,
+			privData, nil, rootCerts); err != nil {
 			return fmt.Errorf("failed to create CA KeyCertBundle (%v)", err)
 		}
-		pkiCaLog.Infof("Using existing public key: %v", string(rootCerts))
+
 	}
 	return err
 }
 
+func loadCacertSecret(client corev1.CoreV1Interface, namespace string, caCertName string, caOpts *IstioCAOptions) error {
```

Comment:
Maybe this change could be separated out into a separate PR? I am concerned about the impact of this change given it is in the critical path and don't want to block other code improvements included in this PR.

Reply:
I can move it - but in normal case it would be a no-op, since if cacerts exist they are mounted. This is mainly to be able to debug Istiod in an IDE - where you can't mount easily ( I have a script to create the various files paths so not a problem to remove this). 

But I can see how this is added yet another option - so I don't mind removing this.

Reply:
This method is still gated on the UseCacertsForSelfSignedCA env var being set. Is it invoked for remote istiod?

Reply:
If the env variable is set - yes. 

It is probably better to remove the gating - one of the comments is to always look up cacerts first, so remote Istiod ( or Istiod in a debugger ) is using them in all cases if user has set it up. 

I'm planning to do that in a separate PR - and after we discuss it, because it would be a behavior change. We probably need to discuss it in the WG - perhaps also remove the UseCacertForSelfSigned ( which is not v1 or enabled by default ) - but I don't want to block this PR which is mostly comments and cleanup on that.